platform/ios/src/MGLScaleBarView.m

```diff
@@ -0,0 +1,289 @@
+    CGFloat padding = CGRectGetMinX(self.frame);
+    return ([self maximumWidth] - padding) / 2;
+}
+
+- (CGFloat)actualWidth {
+    return floorf(self.row.firstObject.floatValue / self.metersPerPoint);
+}
+
+- (CGFloat)minimumWidth {
+    CGFloat fullWidth = CGRectGetWidth(self.superview.bounds);
+    CGFloat padding = CGRectGetMinX(self.frame);
+    CGFloat width = fullWidth - padding;
+    return floorf((width / 2.0) / 3.0);
+}
+
+- (CGFloat)maximumWidth {
+    CGFloat fullWidth = CGRectGetWidth(self.superview.bounds);
+    CGFloat padding = CGRectGetMinX(self.frame);
+    return floorf(fullWidth / 2 - padding);
+}
+
+#pragma mark - Convenient methods
+
+- (NSArray<NSArray<NSNumber*>*> *)validRows {
+    CLLocationDistance minimumDistance = [self minimumWidth] * self.metersPerPoint;
+    CLLocationDistance maximumDistance = [self maximumWidth] * self.metersPerPoint;
+    
+    NSMutableArray *validRows = [NSMutableArray array];
+    for (NSArray<NSNumber *> *row in self.metricTable) {
+        CLLocationDistance distance = row.firstObject.floatValue;
+        if (distance >= minimumDistance && distance <= maximumDistance) {
+            [validRows addObject:row];
+        }
+    }
+    
+    return validRows;
+}
+
+// Find the row closest to -[MGLScaleBarView preferredWidth]
+- (NSArray<NSNumber*> *)preferredRowInRows:(NSArray<NSArray<NSNumber*>*> *)rows {
+    CLLocationDistance diff = CGFLOAT_MAX;
+    CLLocationDistance preferredDistance = [self preferredWidth] * self.metersPerPoint;
+    NSArray<NSNumber *> *preferredRow = rows.firstObject;
+    for (NSArray<NSNumber *> *row in rows) {
+        CLLocationDistance currentDiff = ABS(preferredDistance - row.firstObject.floatValue);
+        if (currentDiff < diff) {
+            preferredRow = row;
+        }
+    }
+    return preferredRow;
+}
+
+- (void)fadeIn {
+    if (self.isVisible) {
+        return;
+    }
+
+    self.visible = YES;
+    
+    [UIView animateWithDuration:.2 delay:0 options:UIViewAnimationOptionBeginFromCurrentState animations:^{
+        self.alpha = 1.0;
+    } completion:nil];
+}
+
+- (void)fadeOut {
+    if (!self.isVisible) {
+        return;
+    }
+    
+    self.visible = NO;
+    
+    [UIView animateWithDuration:.3 delay:0.7 options:UIViewAnimationOptionBeginFromCurrentState animations:^{
+        self.alpha = 0.0;
+    } completion:nil];
+}
+
+#pragma mark - Setters
+
+- (void)setMetersPerPoint:(CLLocationDistance)metersPerPoint {
+    if (_metersPerPoint == metersPerPoint) {
+        return;
+    }
+    
+    [self fadeIn];
+    
+    _metersPerPoint = metersPerPoint;
+    self.row = [self preferredRowInRows:[self validRows]];
+    
+    [self invalidateIntrinsicContentSize];
+    [self setNeedsLayout];
+}
+
+- (void)setRow:(NSArray<NSNumber *> *)row {
+    if (![row isEqualToArray:_row]) {
+        [_bars makeObjectsPerformSelector:@selector(removeFromSuperview)];
+        [_labels makeObjectsPerformSelector:@selector(removeFromSuperview)];
```

Comment:
Would setting `self.subviews` to the empty array also work here, or would that mess up constraints?